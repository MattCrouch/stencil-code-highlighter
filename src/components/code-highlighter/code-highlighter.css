.container {
  font-family: Quicksand, sans-serif;
  margin: 0.5em 0;
  border-radius: 0.25em;
  overflow: hidden;
}

pre[class*="language-"] {
  border-radius: 0;
  margin: 0;
}

.header {
  align-items: center;
  display: flex;
  justify-content: space-between;
  background: #66d9ef;
  padding: 0.5em 1em;
  font-size: 1em;
}

.header span {
  font-weight: bold;
}

button {
  appearance: none;
  background: none;
  border: 0;
  font-family: inherit;
  font-size: inherit;
  cursor: pointer;
  border-radius: 0.25em;
}

button.copy {
  margin: 0 0.25em;
  padding: 0.5em 1em;
  background: rgba(0, 0, 0, 0.1);
}

button.copy:focus, button.copy:hover {
  background: rgba(0, 0, 0, 0.2);
}

.indicator {
  margin-right: 0.5em;
}