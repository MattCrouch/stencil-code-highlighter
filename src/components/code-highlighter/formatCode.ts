import Refractor from "refractor";
import Rehype from "rehype";

export const formatCode = (code: string, type: string) => {
    var nodes = Refractor.highlight(code, type);

    var html = Rehype()
    .stringify({type: 'root', children: nodes})
    .toString();

    return html.trim();
}

export default formatCode;