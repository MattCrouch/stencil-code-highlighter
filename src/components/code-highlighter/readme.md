# code-highlighter

Display highlighted code formatted as if it was in a code editor

<!-- Auto Generated Below -->


## Properties

| Property    | Attribute   | Description | Type      | Default     |
| ----------- | ----------- | ----------- | --------- | ----------- |
| `collapsed` | `collapsed` |             | `boolean` | `false`     |
| `filename`  | `filename`  |             | `string`  | `undefined` |
| `language`  | `language`  |             | `string`  | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
