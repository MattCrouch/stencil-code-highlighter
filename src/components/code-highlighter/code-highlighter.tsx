import { Component, Element, h, Prop, Watch, State } from "@stencil/core";
import formatCode, { isSupported } from "../../utils/formatCode";

@Component({
  tag: "code-highlighter",
  styleUrls: ["prism.css", "code-highlighter.css"],
  shadow: true
})
export class CodeHighlighter {
  @Prop() filename: string;
  @Prop() language: string;
  @Prop({ mutable: true, reflect: true }) collapsed: boolean = false;

  @State() error: boolean = false;

  @Watch("language")
  validateLanguage(value: string) {
    if (typeof value !== "string") {
      this.error = true;
      throw new Error("Language must be set");
    }

    if (!isSupported(value)) {
      this.error = true;
      throw new Error("Language not supported");
    }

    this.error = false;
  }

  @Element() el: HTMLElement;

  connectedCallback() {
    this.validateLanguage(this.language);
  }

  copy() {
    const temp = document.createElement("textarea");
    temp.readOnly = true;
    temp.value = this.el.innerHTML;
    document.body.appendChild(temp);
    temp.select();
    document.execCommand("copy");
    document.body.removeChild(temp);
  }

  toggle() {
    this.collapsed = !this.collapsed;
  }

  render() {
    if (this.error) {
      return null;
    }

    return (
      <div class="container">
        <div class="header">
          <button class="toggle" onClick={this.toggle.bind(this)}>
            <span
              class="indicator"
              innerHTML={this.collapsed ? "&#9656;" : "&#9662;"}
            />
            <span>{this.filename}</span>
          </button>
          <button class="copy" onClick={this.copy.bind(this)}>
            Copy
          </button>
        </div>
        {!this.collapsed && (
          <pre class={`language-${this.language}`}>
            <code
              class={`language-${this.language}`}
              innerHTML={formatCode(this.el.innerHTML, this.language)}
            />
          </pre>
        )}
      </div>
    );
  }
}
