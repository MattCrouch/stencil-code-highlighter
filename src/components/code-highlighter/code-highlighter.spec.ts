// Mock `formatCode` to avoid complicating the unit test
jest.mock("./formatCode", () => (code: string) => code);

// Import Stencil testing tools
import { newSpecPage } from "@stencil/core/testing";

// Import component under test
import { CodeHighlighter } from "./code-highlighter";

// All tests must run async
it("renders as expected", async () => {
  // Generate a SpecPage containing this component
  const component = await newSpecPage({
    components: [CodeHighlighter],
    html: `<code-highlighter language="css"></code-highlighter>`
  });

  // Check it renders as expected
  expect(component.root).toMatchSnapshot();
});
