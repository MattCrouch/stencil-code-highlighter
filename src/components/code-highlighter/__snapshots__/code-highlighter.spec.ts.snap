// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<code-highlighter language="css">
  <mock:shadow-root>
    <div class="container">
      <div class="header">
        <button class="toggle">
          <span class="indicator">
            â–¾
          </span>
          <span></span>
        </button>
        <button class="copy">
          Copy
        </button>
      </div>
      <pre class="language-css">
        <code class="language-css"></code>
      </pre>
    </div>
  </mock:shadow-root>
</code-highlighter>
`;
